File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/XMLElementLiteral.java
/*
*  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
*
*  WSO2 Inc. licenses this file to you under the Apache License,
*  Version 2.0 (the "License"); you may not use this file except
*  in compliance with the License.
*  You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing,
*  software distributed under the License is distributed on an
*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
*  KIND, either express or implied.  See the License for the
*  specific language governing permissions and limitations
*  under the License.
*/
package org.ballerinalang.model.expressions;

import org.ballerinalang.model.NodeLocation;
import org.ballerinalang.model.NodeVisitor;
import org.ballerinalang.model.WhiteSpaceDescriptor;

import java.util.List;
import java.util.Map;

import javax.xml.XMLConstants;

/**
 * {@link XMLElementLiteral} represents an XML literal in Ballerina.
 *
 * @since 0.90
 */
public class XMLElementLiteral extends XMLLiteral {

    private Expression startTagName;
    private Expression endTagName;
    private List<KeyValueExpr> attributes;
    private XMLSequenceLiteral content;
    private XMLElementLiteral parent;
    
    // Namespaces visible to this attribute reference expression.
    private Map<String, Expression> namespaces;
    private Expression defaultNamespaceUri = null;

    public XMLElementLiteral(NodeLocation location, WhiteSpaceDescriptor whiteSpaceDescriptor, Expression tagName) {
        super(location, whiteSpaceDescriptor);
        this.startTagName = tagName;
    }

    public XMLElementLiteral(NodeLocation location, WhiteSpaceDescriptor whiteSpaceDescriptor, Expression tagName,
            List<KeyValueExpr> attributes) {
        this(location, whiteSpaceDescriptor, tagName);
        this.attributes = attributes;
    }

    public Expression getStartTagName() {
        return startTagName;
    }

    public void setStartTagName(Expression startTagName) {
        this.startTagName = startTagName;
    }

    public Expression getEndTagName() {
        return endTagName;
    }

    public void setEndTagName(Expression endTagName) {
        this.endTagName = endTagName;
    }

    public List<KeyValueExpr> getAttributes() {
        return attributes;
    }

    public void setAttributes(List<KeyValueExpr> attributes) {
        this.attributes = attributes;
    }

    public XMLSequenceLiteral getContent() {
        return content;
    }

    public void setContent(XMLSequenceLiteral content) {
        this.content = content;
    }

    public Map<String, Expression> getNamespaces() {
        return namespaces;
    }

    public void setNamespaces(Map<String, Expression> namespaces) {
        Expression defaultNs = namespaces.remove(XMLConstants.DEFAULT_NS_PREFIX);
        if (defaultNs != null) {
            defaultNamespaceUri = defaultNs;
        }
        this.namespaces = namespaces;
    }

    public Expression getDefaultNamespaceUri() {
        return defaultNamespaceUri;
    }

    public void setDefaultNamespaceUri(Expression defaultNamespaceUri) {
        this.defaultNamespaceUri = defaultNamespaceUri;
    }

    public XMLElementLiteral getParent() {
        return parent;
    }

    public void setParent(XMLElementLiteral parent) {
        this.parent = parent;
    }

    @Override
    public void accept(NodeVisitor visitor) {
        visitor.visit(this);
    }
}

File: modules/web/js/ballerina/ast/expressions/xml-element-literal.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import _ from 'lodash';
import Expression from './expression';
import ASTFactory from './../ast-factory';

/**
 * Class for XML Element Literal.
 * @class XMLElementLiteral.
 * */
class XMLElementLiteral extends Expression {
    constructor(args) {
        super('XMLElementLiteral');
        this._attributes = _.get(args, 'attributes', []);
        this.type_name = _.get(args, 'type_name', '');
    }

    getExpressionString(isTemplate) {
        let expression = '';
        expression += isTemplate ? '' : `${this.type_name} \``;
        if (this.children[0]) {
            if (ASTFactory.isBasicLiteralExpression(this.children[0])) {
                expression += `<${this.children[0].getBasicLiteralValue()}`;
            } else if (ASTFactory.isSimpleVariableReferenceExpression(this.children[0])) {
                expression += `<{{${this.children[0].getVariableName()}}}`;
            } else if (ASTFactory.isXMLQNameExpression(this.children[0])) {
                expression += `<${this.children[0]._localName}`;
            }

            this._attributes.forEach((attribute) => {
                attribute.children.forEach((child) => {
                    if (ASTFactory.isBasicLiteralExpression(child)) {
                        if (attribute.children.indexOf(child) === 0 || attribute.children.indexOf(child) === 1) {
                            expression += `="${child.getBasicLiteralValue()}`;
                        } else {
                            expression += `${child.getBasicLiteralValue()}`;
                        }
                        if (attribute.children.indexOf(child) === (attribute.children.length - 1)) {
                            expression += '"';
                        }
                    } else if (ASTFactory.isSimpleVariableReferenceExpression(child)) {
                        if (attribute.children.indexOf(child) === 0) {
                            expression += ` {{${child.getVariableName()}}}`;
                        } else if (attribute.children.indexOf(child) === 1) {
                            expression += `="{{${child.getVariableName()}}}`;
                        } else {
                            expression += `{{${child.getVariableName()}}}`;
                        }

                        if (attribute.children.indexOf(child) !== 0
                            && attribute.children.indexOf(child) === (attribute.children.length - 1)) {
                            expression += '"';
                        }
                    } else if (ASTFactory.isXMLQNameExpression(child)) {
                        expression += ` ${child.getLocalName()}`;
                    } else if (ASTFactory.isBinaryExpression(child)) {
                        expression += `="${child.getExpressionString(true)}"`;
                    }
                });
            });
            if (!this.children[1] && !this.children[2]) {
                expression += '/>';
            } else {
                expression += '>';
            }
        }

        if (ASTFactory.isXMLSequenceLiteral(this.children[1]) && this.children[1].children.length > 0) {
            expression += this.children[1].getExpressionString(true);
        } else if (ASTFactory.isXMLQNameExpression(this.children[1])) {
            if (ASTFactory.isBasicLiteralExpression(this.children[1])) {
                expression += `</${this.children[1]._localName}>`;
            } else if (ASTFactory.isSimpleVariableReferenceExpression(this.children[1])) {
                expression += `</{{${this.children[1].getVariableName()}}}>`;
            } else {
                expression += `</${this.children[1]._localName}>`;
            }
        }

        if (this.children[2]) {
            if (ASTFactory.isBasicLiteralExpression(this.children[2])) {
                expression += `</${this.children[2]._localName}>`;
            } else if (ASTFactory.isSimpleVariableReferenceExpression(this.children[2])) {
                expression += `</{{${this.children[2].getVariableName()}}}>`;
            } else {
                expression += `</${this.children[2]._localName}>`;
            }
        }

        expression += isTemplate ? '' : '`';
        return expression;
    }

    /**
     * Initialize the model from the json Node.
     * @param {Object} jsonNode - Json node fore XML Element Literal.
     * */
    initFromJson(jsonNode) {
        this.children = [];
        this.type_name = _.get(jsonNode, 'type_name', '');
        this._attributes = [];
        _.forEach(jsonNode.children, (childNode) => {
            const child = ASTFactory.createFromJson(childNode);
            this.addChild(child, undefined, true, true);
            child.initFromJson(childNode);
        });

        _.forEach(jsonNode.attributes, (childNode) => {
            const child = ASTFactory.createFromJson(childNode);
            this._attributes.push(child);
            child.initFromJson(childNode);
        });
    }
}

export default XMLElementLiteral;
